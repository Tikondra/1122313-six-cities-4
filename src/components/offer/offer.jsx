import React from "react";
import PropTypes from "prop-types";
import {MapType, PlacesListClass, TypePlace} from "../../constants";
import OfferGallery from "../offer-gallery/offer-gallery.jsx";
import {getRating} from "../../utils";
import Options from "../options/options.jsx";
import Host from "../host/host.jsx";
import Reviews from "../reviews/reviews.jsx";
import Map from "../map/map.jsx";
import PlacesList from "../places-list/places-list.jsx";

const getPremium = (isPremium) => isPremium ?
  <div className="property__mark">
    <span>Premium</span>
  </div> :
  ``;

const Offer = ({offer, offers, onClickByHeader, onHoverPlace}) => {
  const {
    description,
    guests,
    host,
    isPremium,
    options,
    pictures,
    price,
    rating,
    room,
    title,
    type,
    reviews,
  } = offer;

  const nearbyPlaces = offers.slice(0, 3); // временно, пока нет поиска поблизости

  return (
    <main className="page__main page__main--property">
      <section className="property">
        <OfferGallery
          pictures = {pictures}
          title = {title}
        />
        <div className="property__container container">
          <div className="property__wrapper">
            {getPremium(isPremium)}
            <div className="property__name-wrapper">
              <h1 className="property__name">
                {title}
              </h1>
              <button className="property__bookmark-button button" type="button">
                <svg className="property__bookmark-icon" width="31" height="33">
                  <use xlinkHref="#icon-bookmark"/>
                </svg>
                <span className="visually-hidden">To bookmarks</span>
              </button>
            </div>
            <div className="property__rating rating">
              <div className="property__stars rating__stars">
                <span style={{width: getRating(rating)}}/>
                <span className="visually-hidden">Rating</span>
              </div>
              <span className="property__rating-value rating__value">{rating}</span>
            </div>
            <ul className="property__features">
              <li className="property__feature property__feature--entire">
                {type}
              </li>
              <li className="property__feature property__feature--bedrooms">
                {room} Bedrooms
              </li>
              <li className="property__feature property__feature--adults">
                Max {guests} adults
              </li>
            </ul>
            <div className="property__price">
              <b className="property__price-value">&euro;{price}</b>
              <span className="property__price-text">&nbsp;night</span>
            </div>
            <Options options={options}/>
            <Host
              host = {host}
              description = {description}
            />
            <Reviews
              reviews={reviews}
            />
          </div>
        </div>
        <Map
          type = {MapType.PROPERTY}
          offers = {offers}
        />
      </section>
      <div className="container">
        <section className="near-places places">
          <h2 className="near-places__title">Other places in the neighbourhood</h2>
          <PlacesList
            className = {PlacesListClass.PROPERTY}
            offers={nearbyPlaces}
            onClickByHeader={onClickByHeader}
            onHoverPlace={onHoverPlace}
          />
        </section>
      </div>
    </main>
  );
};

Offer.propTypes = {
  offer: PropTypes.shape({
    isPremium: PropTypes.bool.isRequired,
    price: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    type: PropTypes.oneOf(Object.values(TypePlace)).isRequired,
    rating: PropTypes.number.isRequired,
    pictures: PropTypes.array.isRequired,
    description: PropTypes.array.isRequired,
    room: PropTypes.number.isRequired,
    guests: PropTypes.number.isRequired,
    options: PropTypes.array.isRequired,
    host: PropTypes.shape({
      avatar: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      isSuper: PropTypes.bool.isRequired,
    }).isRequired,
    reviews: PropTypes.array,
  }),
  offers: PropTypes.array.isRequired,
  onClickByHeader: PropTypes.func.isRequired,
  onHoverPlace: PropTypes.func.isRequired,
};

export default Offer;
