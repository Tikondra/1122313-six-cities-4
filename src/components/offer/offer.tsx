import * as React from "react";
import {connect} from "react-redux";
import {withRouter, Redirect} from 'react-router-dom';
import {Format, MapType, PlacesListClass} from "../../constants";
import OfferGallery from "../offer-gallery/offer-gallery";
import {getRating} from "../../utils";
import Options from "../options/options";
import Host from "../host/host";
import Reviews from "../reviews/reviews";
import Map from "../map/map";
import PlacesList from "../places-list/places-list";
import {getAuthorizationStatus} from "../../reducer/user/selectors";
import {getNearbyPlaces, getPlaceForId, getReviews} from "../../reducer/data/selectors";
import {getCity} from "../../reducer/app-state/selectors";
import Header from "../header/header";
import FavoriteBtn from "../favorite-btn/favorite-btn";
import {AuthStatus, Offer as TypeOffer} from "../../types";
import {Operation} from "../../reducer/data/data";
import {ActionCreator} from "../../reducer/app-state/app-state";

interface Props {
  offer: TypeOffer;
  activeCity: {
    coordinates: [];
  };
  onClickByHeader: () => void;
  onHoverPlace: () => void;
  onClickByFavorite: () => void;
  status: AuthStatus.AUTH | AuthStatus.NO_AUTH;
  reviews: [];
  nearbyPlaces: TypeOffer[];
}

const getPremium = (isPremium) => isPremium ?
  <div className="property__mark">
    <span>Premium</span>
  </div> :
  ``;

const Offer: React.FC<Props> = ({
  offer, activeCity, status, reviews = [], nearbyPlaces = [],
  onClickByHeader, onHoverPlace, onClickByFavorite}: Props) => {

  if (!offer) {
    return <Redirect to="/" />;
  }

  const {description, guests, host, isPremium, isFavorite, options, pictures, price, rating, room, title, type} = offer;

  return (
    <div className="page page--gray page--property">
      <Header/>
      <main className="page__main page__main--property">
        <section className="property">
          <OfferGallery
            pictures = {pictures}
            title = {title}
          />
          <div className="property__container container">
            <div className="property__wrapper">
              {getPremium(isPremium)}
              <div className="property__name-wrapper">
                <h1 className="property__name">
                  {title}
                </h1>
                <FavoriteBtn
                  status={status}
                  size={Format.BOOKMARK_SIZE.BIG}
                  type={Format.BOOKMARK_TYPE.OFFER}
                  isActive={isFavorite}
                  onClickByFavorite = {onClickByFavorite}
                />
              </div>
              <div className="property__rating rating">
                <div className="property__stars rating__stars">
                  <span style={{width: getRating(rating)}}/>
                  <span className="visually-hidden">Rating</span>
                </div>
                <span className="property__rating-value rating__value">{rating}</span>
              </div>
              <ul className="property__features">
                <li className="property__feature property__feature--entire">
                  {type}
                </li>
                <li className="property__feature property__feature--bedrooms">
                  {room} Bedrooms
                </li>
                <li className="property__feature property__feature--adults">
                Max {guests} adults
                </li>
              </ul>
              <div className="property__price">
                <b className="property__price-value">&euro;{price}</b>
                <span className="property__price-text">&nbsp;night</span>
              </div>
              <Options options={options}/>
              <Host
                host = {host}
                description = {description}
              />
              <Reviews
                reviews={reviews}
                status = {status}
              />
            </div>
          </div>
          <Map
            type = {MapType.PROPERTY}
            offers = {nearbyPlaces}
            activeOffer={offer}
            city = {activeCity}
          />
        </section>
        <div className="container">
          <section className="near-places places">
            <h2 className="near-places__title">Other places in the neighbourhood</h2>
            <PlacesList
              status={status}
              className = {PlacesListClass.PROPERTY}
              classNameCard={PlacesListClass.PROPERTY_ARTICLE_NAME}
              classCard = {PlacesListClass.PROPERTY_CARD}
              imgSize = {Format.PLACE_IMG_SIZE.NORMAL}
              offers={nearbyPlaces}
              onClickByHeader={onClickByHeader}
              onHoverPlace={onHoverPlace}
              onClickByFavorite = {onClickByFavorite}
            />
          </section>
        </div>
      </main>
    </div>
  );
};

const mapStateToProps = (state, {match}) => ({
  status: getAuthorizationStatus(state),
  activeCity: getCity(state),
  offer: getPlaceForId(state, match.params.id),
  reviews: getReviews(state),
  nearbyPlaces: getNearbyPlaces(state),
});

const mapDispatchToProps = (dispatch) => ({
  onHoverPlace(offer) {
    dispatch(ActionCreator.changeActiveOffer(offer));
  },

  onClickByHeader(id) {
    dispatch(Operation.loadReviews(id));
    dispatch(Operation.loadNearbyPlaces(id));
  },

  onClickByFavorite() {
    dispatch(Operation.postFavorite());
  }
});

export {Offer};
export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Offer));
