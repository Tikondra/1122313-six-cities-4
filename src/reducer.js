import offers from "./mocks/offers";
import {Cities} from "./mocks/const";
import {extend} from "./utils";
import {PageType} from "./constants";

const initialState = {
  city: Cities[0],
  offers,
  cities: Cities,
  page: PageType.MAIN,
  activeOffer: null,
};

const ActionType = {
  CHANGE_CITY: `CHANGE_CITY`,
  GET_OFFERS: `GET_OFFERS`,
  CHANGE_OFFER: `CHANGE_OFFER`,
  CHANGE_PAGE: `CHANGE_PAGE`,
};

const ActionCreator = {
  changeActiveOffer: (offer) => ({
    type: ActionType.CHANGE_OFFER,
    payload: offer,
  }),

  changePage: () => ({
    type: ActionType.CHANGE_PAGE,
    payload: PageType.PROPERTY,
  }),

  getOffers: () => ({
    type: ActionType.CHANGE_CITY,
    payload: 1,
  }),

  changeCity: () => ({
    type: ActionType.CHANGE_CITY,
    payload: 1,
  }),
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_CITY:
      return extend(state, {
        city: action.payload,
      });

    case ActionType.CHANGE_OFFER:
      return extend(state, {
        activeOffer: action.payload,
      });

    case ActionType.CHANGE_PAGE:
      return extend(state, {
        page: action.payload,
      });

    case ActionType.GET_OFFERS:
      return extend(state, {
        offers: action.payload,
      });
  }

  return state;
};

export {reducer, ActionType, ActionCreator};
