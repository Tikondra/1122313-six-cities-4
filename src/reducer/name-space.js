const NameSpace = {
  DATA: `DATA`,
  APP_STATE: `APP_STATE`,
  USER: `USER`,
  REVIEW: `REVIEW`
};

export default NameSpace;
