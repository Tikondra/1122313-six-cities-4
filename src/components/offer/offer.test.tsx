import * as React from "react";
import {createStore} from 'redux';
import {Provider} from "react-redux";
import * as renderer from "react-test-renderer";
import {Offer} from "./offer";
import {BrowserRouter, Route} from "react-router-dom";
import reducer from "../../reducer/reducer";

const store = createStore(reducer);

it(`should render Offer`, function () {
  const tree = renderer
    .create(
        <BrowserRouter>
          <Provider store={store}>
            <Route component={Offer}/>
          </Provider>
        </BrowserRouter>
    ).toJSON();

  expect(tree).toMatchSnapshot();
});
