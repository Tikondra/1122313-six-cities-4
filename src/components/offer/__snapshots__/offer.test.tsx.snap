// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Offer 1`] = `null`;
